### ISSUE TITLE
submitted chats should show immediately in the list (before server poll)

### ISSUE THREAD

**Issue by seanmakesgames** [DrizzlyBear/hackmud_chat_web#4]
this is better feedback, and will mask network connectivity

question is, do we reorder messages that happened before you sent yours (and hadn't received yet), or do we leave as is?

**Comment by ethankaminski** [comment]
Something that came up in discussion with dtr about this: It would be convenient if the `create_chat` endpoint returned the same response body as the `chats` endpoint. Namely, if you're chatting in a channel, you probably want to poll *anyway*; getting a list of available messages, after you send one of yours, would be a useful thing for client writers.

**Comment by seanmakesgames** [comment]
yes, that's reasonable, but you'll still probably have a poll already in flight- and while your chat is being sent to the 'create_chat' endpoint we still want to have feedback for the user

**Comment by seanmakesgames** [comment]
![default](https://cloud.githubusercontent.com/assets/568152/23675419/23dccd9a-032e-11e7-90d4-62fbd0b9cf34.png)

**Comment by KuroTsuto** [comment]
IMO, I'd like to the see submitted messages added either to the bottom of the message list - or perhaps a separate "pending" list just above the chat input - and kept there until the server confirms successful receipt (at which point they would appear in the message list at their proper location). Pending messages would have some sort of distinct identifying trait - maybe partial opacity or some sort of loading spinner (the animated `>`s used in-game, possibly).

**Comment by seanmakesgames** [comment]
sounds good. we could probably work something out like that

**Comment by hackmud-dtr** [comment]
One problem I have with this is a thing that we see on dsicord a lot: it means that people will see messages in a different order. I.e. sender sees their message immediately, but everyone else sees it after a message someone else sent (that sender hasn't seen first); to me, consistency (i.e. there is one true order of messages and everyone sees the one true order) is far more important than a 2-second lag on showing a message, but maybe I'm lone in that.

Also this implies showing the message before we even get a confirmation back from the API if the message was sent (it might not be, due to rate limiting), and then what... delete it, I guess? 

Also, if a message is sent close to a minute boundary (say, 59.75s), we might insert it with a timestamp like XX45 but then the server doesn't see it til after the minute mark and so to everyone else it'll display as XX46. Basically, without knowing the details that we can't know until after it's gotten to the server, displaying it early has a ton of chances to be wrong, and I'd rather be right slowly than wrong quickly. 



We COULD try showing it early, then re-ordering it and changing the date, etc, after the next poll, but that adds a lot of overhead. And since we don't have ids yet we can't even map to an exact message, so we have to check content (again: times can change so we can't rely on those). Gets messy fast.

**Comment by seanmakesgames** [comment]
yes it's complex. yes it's worth it.
We can spec it out together and figure out which parts are important to implement and which can be held off on.
Top of head:
show the message immediately, show retry on send failure (due to rate limit or whatever other reason), use send api response information (this may not exist yet) to fix location in time / ordering etc.

likely we'll want to also put some rate limit estimates on client-side, or provide more data back on the server response to provide 'next message allowed at' if that doesn't exist.

**Comment by hackmud-dtr** [comment]
Yeah, right now the server really just says "sorry, rate limited", it doesn't say when to retry (unless it does in a retry-after header, which would make sense but I totally ignore headers when dealing with JSON usually). Having some concept of when next message can be sent (also in-game!) would be great.

Another thing to add to the response: the id of the new chat. I.e. create_chat should return the id of that chat once it is made.

Some complications on the reorder-and-update idea to consider (I'll try to work on a solution tonight, i just want to get notes down)

* multiple identical-text messages sent in a row (someone hammering "test" while debugging a flaky internet connection; i.e. same-message-content won't be good enough to determine what message is what, we have to actually track the responses attached to each specific message)

* Server might get messages sent by a client out of order. i.e. 1 client sends messages A, B, and C in that order, rapid fire. But because of network oddities, server order winds up being B, C, A (i.e. A got stuck somewhere for a long enough for B and C to get to the server ahead of it). So be ready to re-order our own messages relative to each other. 

* Server might implement transforms (i.e. in the client, :::TRUST COMMUNICATION::: gets corrupted, but not currently in the chat API, but I could see in the future the chat API doing that instead of the client), so can't even rely on message content not needing to be changed. 



Anyway i think best bet right now is this:
As soon as you hit enter, message is inserted but at like 50% opacity to show it's pending. Internally we track the response and when the response comes back as accepted, we mark that as sent and change it to say 75% opacity. It isn't quite done yet. Then, the next time the client polls, we update to put that message in it's correct place and correct times/contents based on the server response, and finally it gets 100% opacity (this is why create_chat should return the id: that way we can tell which message in the server response matches up to the chat we just made, by id)

**Comment by seanmakesgames** [comment]
create_chat will provide the id, all you've got to do is hold onto the js chat object that was created with the request and you can update them in the callback in whatever order they are processed by the server. server response can also contain the timestamp for the message and the 'final' content, too. Wait to implement this until the api updates happen that make it easier to impl.

**Comment by KuroTsuto** [comment]
I guess my question at this juncture is if we're aiming to enhancing the chat experience, or imitate the behavior and interface of the game. If the former (which other issues like #23 might allude to), then positioning pending messages based on when you *sent* them makes a lot of sense.

But if the latter, keeping pending messages in a list separate from channel messages simplifies things significantly:

- Differentiating identical messages becomes irrelevant; you can just enumerate them
- Polled messages remain the authority on message order; whenever you sent or received your sent message, all users still see the messages in the same order. Admittedly, a means to assert when the message was sent would be neat, but I don't believe the client currently does this(?)
- Error handling is much easier: splice out the pending message corresponding to the failed request and insert an error in either channel messages or in-place in the pending list, perhaps with a link to retry depending on the error.

I built a simple proof-of-concept implementation of the separate list approach a week or two ago but never submitted a PR because I didn't like where I was stapling all of the code on. I think the most complicated part of the code was my crappy `>>>>>>>>` animation. Server-side message transformation would still kill it though.

In any scenario, the discussed API changes would definitely be dope.
